eth/fetcher/tx_fetcher.go

```diff
@@ -0,0 +1,289 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package fetcher
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	txAnnounceLimit = 4096            // Maximum number of unique transaction a peer may have announced
+	txFetchTimeout  = 5 * time.Second // Maximum allotted time to return an explicitly requested transaction
+
+	// MaxTransactionFetch is the maximum transaction number can be fetched
+	// in one request. The rationale to pick this value is:
+	// In eth protocol, the softResponseLimit is 2MB. While the max size of
+	// valid transaction is 128KB right now. Worst case scenario, a packet
+	// can contain 16 transactions, but normally 2MB is totally enough for
+	// containing 64 transactions.
+	MaxTransactionFetch = 64
```

Comment:
I think we should bump this up. If we overshoot, the extra hashes are wasted, but since a hash is tiny, the cost is negligible. However, the "cost" of extra network RTTs can be impactful if we need to do many requests.

Looking at a few recent blocks on Etherscan, they seem to contain transactions weighing <200B on average. If we'd like to have meaningful reply messages of 128KB, that's 640 transactions of this size. Maybe that's a bit high to waste if a few txs are large, but let's try a middle ground. Let's set this limit to 256. That should allow packing up 50KB of useful data if tx are small and wastes 8KB if txs are gigantic (1 per reply).

Reply:
Then again, I can be convinced otherwise too. CC @holiman @AlexeyAkhunov ?